Make unstruct resolvable and introduce `root.unwrap(TgpuVertexLayout)` (#781)
software-mansion/TypeGPU: apps/typegpu-docs/src/content/docs/fundamentals/vertex-layouts.mdx

---
title: Vertex Layouts
description: A guide on how to create and use typed vertex layouts
draft: true
---

> Typed vertex layouts are a way to describe the structure of vertex data in a typed manner. They are used to create a single source of truth for vertex data, which can be used in combination with [`tgpu.resolve`](/TypeGPU/fundamentals/resolve) and [`root.unwrap`](/TypeGPU/fundamentals/roots) to easily create and manage vertex buffers.

Reviewer: Imo, too many **types** 🧈🧈. I am wondering if we can reword it to showcase its purpose without relating to its WebGPU counterpart. Those who know WebGPU will still be able to see the improvements, and those who don't will be able to enjoy the docs and follow along 📖

```suggestion
Vertex layouts are much like bind group layouts, in that they define the relationship between shaders and buffers. More precisely, they define what vertex attributes a shader expects, and how they are laid out in the corresponding vertex buffer.
```

## Creating a vertex layout

> To create a vertex layout, use the `tgpu.vertexLayout` function. It takes a function that takes a number and returns an array containing any data type, and a string that describes the type of the vertex layout. It can be either `vertex` or `instance` (default is `vertex`).

Reviewer: ```suggestion
To create a vertex layout, use the `tgpu.vertexLayout` function. It takes an array schema constructor, i.e., a function that returns an array schema given the number of elements to render (vertices/instances). To determine what each element of the array corresponds to, you can pass an optional `stepMode` argument, which can be either `vertex` (default) or `instance`.
```

```ts
import tgpu from 'typegpu';
import * as d from 'typegpu/data';

const ParticleGeometry = d.struct({
  tilt: d.f32,
  angle: d.f32,
  color: d.vec4f,
});

const geometryLayout = tgpu

>   .vertexLayout((n: number) => d.arrayOf(ParticleGeometry, n), 'instance');

Reviewer: ```suggestion
  .vertexLayout((n) => d.arrayOf(ParticleGeometry, n), 'instance');
```

- <a href="https://developer.mozilla.org/en-US/docs/Web/API/GPURenderPassEncoder" target="_blank" rel="noopener noreferrer">Render pass encoding</a>
:::

Vertex layouts are much like bind group layouts, in that they define the relationship between shaders and buffers. More precisely, they define what vertex attributes a shader expects, and how they are laid out in the corresponding vertex buffer.

## Creating a vertex layout

To create a vertex layout, use the `tgpu.vertexLayout` function. It takes an array schema constructor, i.e., a function that returns an array schema given the number of elements to render (vertices/instances). To determine what each element of the array corresponds to, you can pass an optional `stepMode` argument, which can be either `vertex` (default) or `instance`.

```ts
import tgpu from 'typegpu';
import * as d from 'typegpu/data';

const ParticleGeometry = d.struct({
  tilt: d.f32,
  angle: d.f32,
  color: d.vec4f,
});

const geometryLayout = tgpu
  .vertexLayout((n) => d.arrayOf(ParticleGeometry, n), 'instance');
```

## Utilizing loose schemas with vertex layouts

> The above example is great if the vertex buffer will also be used as a storage or uniform buffer. However, if you're only interested in the vertex data, you can use a loose schema instead.
> It is not restricted by alignment rules and can utilize many useful [vertex formats](https://www.w3.org/TR/webgpu/#vertex-formats). To create loose schemas, use `d.unstruct` instead of `d.struct` and `d.disarrayOf` instead of `d.arrayOf`.
> Inside of loose schemas you can use vertex formats as well as the usual data types.

Reviewer: Thank you ChatGPU 🫡 
```suggestion
If the vertex buffer is not required to function as a storage or uniform buffer, a *loose schema* may be used to define the vertex data layout. Loose schemas are not subject to alignment restrictions and allow the use of various [vertex formats](https://www.w3.org/TR/webgpu/#vertex-formats).

To define a loose schema:
- Use `d.unstruct` instead of `d.struct`.
- Use `d.disarrayOf` instead of `d.arrayOf`.

Within a loose schema, both standard data types and vertex formats can be utilized.
```

```ts
const LooseParticleGeometry = d.unstruct({
  tilt: d.f32,
  angle: d.f32,

>   color: d.unorm8x4, // 4x8-bit unsigned normalized

Reviewer: ```suggestion
  // four 8-bit values, unsigned & normalized
  // i.e., four integers in (0, 255) represent four floats in the range of (0.0, 1.0)
  color: d.unorm8x4,
```

//      "offset": 0,
//      "shaderLocation": 0
//    },
//    {
//      "format": "float32",
//      "offset": 4,
//      "shaderLocation": 1
//    },
//    {
//      "format": "float32x4",
//      "offset": 16,
//      "shaderLocation": 2
//    }
//  ]
//}
```

This will return a `GPUVertexBufferLayout` that can be used when creating a render pipeline.

```diff lang=ts
const renderPipeline = device.createRenderPipeline({
  layout: device.createPipelineLayout({
    bindGroupLayouts: [root.unwrap(bindGroupLayout)],
  }),
  primitive: {
    topology: 'triangle-strip',
  },
  vertex: {
    module: renderShader,
-    buffers: [
-      {
-        arrayStride: 32,
-        stepMode: 'instance',
-        attributes: [
-          {
-            format: 'float32',
-            offset: 0,
-            shaderLocation: 0,
-          },
-          {
-            format: 'float32',
-            offset: 4,
-            shaderLocation: 1,
-          },
-          {
-            format: 'float32x4',
-            offset: 16,
-            shaderLocation: 2,
-          },
-        ],
-      },
-    ],
+    buffers: [root.unwrap(geometryLayout)],
  },
  fragment: {
    ...
  },
});
```

> If you are using a loose schema, you can now resolve it to get its WGSL representation.

Reviewer: ```suggestion
Loose schemas can be interpreted in multiple ways within a shader. However, for convenience, they can be resolved to their default WGSL representation.
```
